source/fundamentals/bson.txt

```diff
@@ -0,0 +1,231 @@
+             FirstName string `bson:"first_name,omitempty"`
+             LastName  string `bson:"last_name,omitempty"`
+             Age       int    `bson:"age,omitempty"`
+         }
+         ...
+
+         coll := client.Database("school").Collection("students")
+         student1 := Student{ FirstName : "Arthur", Age : 8}
+         insertResult, err := coll.InsertOne(context.TODO(), student1)
+
+      The corresponding BSON representation looks like this:
+
+      .. code-block:: json
+         :copyable: false
+      
+         { "_id" : ObjectId("6101c197a5284b5dfcbfd90b"), "first_name" : "Arthur", "age" : 8 }
+
+      In this example, including struct tags lets the driver set custom BSON
+      field names such as ``first_name`` and omit the empty ``LastName`` field during
+      marshalling.
+
+      .. tabs::
+
+   .. tab:: No Struct Tags
+      :tabid: no-struct-tags
+
+      The following example demonstrates how the Go Driver marshals a
+      struct without any struct tags:
+
+      .. code-block:: go
+
+         type Student struct {
+             FirstName string
+             LastName  string
+             Age       int
+         }
+         ...
+
+         coll := client.Database("school").Collection("students")
+         student1 := Student{ FirstName : "Arthur", Age : 8}
+         insertResult, err := coll.InsertOne(context.TODO(), student1)
+
+      The corresponding BSON representation looks like this:
+
+      .. code-block:: json
+         :copyable: false
+      
+         { "_id" : ObjectId("6101c197a5284b5dfcbfd90b"), "firstname" : "Arthur", "lastname" : "", "age" : 8 }
+         
+      In this example, the Go Driver automatically sets the lowercase
+      of the struct fields as the BSON field names and includes an empty
+      ``lastname`` field in the BSON representation.
+
+
+Unmarshalling
+-------------
+
+You can unmarshal BSON documents using the ``Decode()`` method. When you
+execute either a ``FindOne()`` or ``Find()`` query, you can use the
+``Decode()`` method on the return object to unmarshal the results.
+
+The ``Decode()`` method returns an ``error`` interface type which
+contains one of the following values:
+
+- ``nil`` if a document matched your query, and there were no errors retrieving
+  the document. 
+- If the driver retrieved your document but could not unmarshal your result, the
+  ``Decode()`` method returns the unmarshalling error.
+- If there was an error retrieving your document during execution of the
+  ``FindOne()`` method, the error propagates to the ``Decode()`` method and
+  the ``Decode()`` method returns the error. 
+
+When used on the ``SingleResult`` type returned by the ``FindOne()``
+method, ``Decode()`` can also return the ``ErrNoDocuments`` error if no
+documents matched the query filter.
+
+The following example demonstrates how you can use the ``Decode()``
+method to unmarshal and read the result of a simple ``FindOne()``
+operation:
+
+.. code-block:: go
```

Comment:
Would it be more informative to users here to show unmarshalling into a user-defined type (i.e. the `Student` type defined above) rather than `bson.D`? I'm not super familiar with the Go API, but in Rust, we return the `Document` type from our `find_one` method, so the Serde API is mostly useful to users when they're deserializing results back into their custom types.